### PR TITLE
OCPBUGSM-27367 Cancel controller when installation is cancelled

### PR DESCRIPTION
**Pull request by slaviered**
- Indication of log collection complete is sent on this case too
  - Save resources and irrelevant setup work on abandoned clusters

### CODE REVIEW

File: src/assisted_installer_controller/assisted_installer_controller.go [hunk 1]
```diff
@@ -127,7 +137,17 @@ func (c *controller) WaitAndUpdateNodesStatus(status *ControllerStatus) {
 
 		hostsInProgressMap := common.GetHostsInStatus(assistedNodesMap, []string{models.ClusterStatusInstalled}, false)
 		errNodesMap := common.GetHostsInStatus(hostsInProgressMap, []string{models.HostStatusError}, true)
+		cancelNodesMap := common.GetHostsInStatus(hostsInProgressMap, []string{models.HostStatusCancelled}, true)
```
Comment (tsorya) [inline src/assisted_installer_controller/assisted_installer_controller.go:140]: can't we use the same map as errors? why not to add cancelled status to https://github.com/openshift/assisted-installer/pull/270/files#diff-1da579fda3bc689940661d48dd744a46f23824270427207cfae4f8c7448eddd2R139
?

Comment (slaviered) [reply src/assisted_installer_controller/assisted_installer_controller.go:140]: It's possible under the assumption that cancel and error behaves the same (all hosts are moved to that status once it detected in the service)

Comment (tsorya) [reply src/assisted_installer_controller/assisted_installer_controller.go:140]: Can we validate it? I think it is same behavior

Comment (slaviered) [reply src/assisted_installer_controller/assisted_installer_controller.go:140]: Yes, I validated it and all host transit to either cancel or error. However, in the controller there is no unified behavior for cancel vs error: on error it tried to collect must-gather and on cancel it does not. This is why we need to have separate indication for cancel vs error

File: src/assisted_installer_controller/assisted_installer_controller.go [hunk 2]
```diff
@@ -889,8 +909,8 @@ func (c *controller) UploadLogs(ctx context.Context, cancellog context.CancelFun
 				podName = pods[0].Name
 			}
 
-			if status.HasError() {
-				c.log.Infof("Error detected. Closing logs and aborting...")
+			if status.HasError() || status.Cancelled() {
```
Comment (tsorya) [inline src/assisted_installer_controller/assisted_installer_controller.go:912]: Maybe i forgot something but i don't remember why we have logic here? 
In any case i don't like the idea of adding new status. If this is needed somehow (i truly believe we can cancel logs from outside) i prefer to make status struct to handle cancel and error case together.

Comment (tsorya) [reply src/assisted_installer_controller/assisted_installer_controller.go:912]: why we can't just disable this statement 
https://github.com/openshift/assisted-installer/blob/master/src/main/assisted-installer-controller/assisted_installer_main.go#L84 ?

Comment (slaviered) [reply src/assisted_installer_controller/assisted_installer_controller.go]: After discussion I am implementing a mechanism to cancel all the go routines on cancel and error

File: src/installer/installer.go [hunk 3]
```diff
@@ -346,13 +346,14 @@ func (i *installer) downloadHostIgnition() (string, error) {
 }
 
 func (i *installer) waitForNetworkType(kc k8s_client.K8SClient) error {
-	return utils.WaitForPredicate(waitForeverTimeout, 5*time.Second, func() bool {
+	err, _, _ := Wait().WithNoTimeout().WithInterval(5 * time.Second).WithPredicate(func() bool {
```
Comment (tsorya) [inline src/installer/installer.go:349]: why without timeout?

File: src/utils/waiting/waiting.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+		interval: GeneralWaitInterval,
+	}
+}
+
+func (p *wait) WithPredicate(predicate func() bool) *wait {
+	p.predicate = predicate
+	return p
+}
+
+func (p *wait) WithTimeout(timeout time.Duration) *wait {
```
Comment (tsorya) [inline src/utils/waiting/waiting.go:36]: where it is used?

File: src/utils/waiting/waiting.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+	p.cancelctx = cancelctx
+	return p
+}
+
+func (p *wait) WithMessage(message string) *wait {
+	p.message = message
+	return p
+}
+
+func (p *wait) Start() (error, error, bool) {
```
Comment (tsorya) [inline src/utils/waiting/waiting.go:61]: why 2 errors are needed? why does it matter if it was canceled or not? 
Cancel is coming from outside so what's point of verifying error if i just cancelled it by myself?

File: src/utils/waiting/waiting.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
+package waiting
```
Comment (tsorya) [inline src/utils/waiting/waiting.go:1]: I must say that i don't really understand what this package gives us. It sounds like all we need one function and transformed it to object that still only one function from it is used? maybe i miss the point?

File: src/installer/installer.go (hunk 3 above)
Comment (tsorya) [reply src/installer/installer.go:349]: i am not sure i really like the usage i must say. In any case i think it should start with WithPredicate and then adding with interval and etc. 
maybe we can just move to https://github.com/avast/retry-go ?

File: src/main/assisted-installer-controller/assisted_installer_main.go [hunk 7]
```diff
@@ -55,36 +55,48 @@ func main() {
...
 	var wgLogs sync.WaitGroup
-	var status assistedinstallercontroller.ControllerStatus
 
-	ctxApprove, cancelApprove := context.WithCancel(context.Background())
+	//every context should be added to the watch list before calling
+	//the routines because they may be canceled within those routines
+	//espceially, they should be called before WaitAndUpdateNodesStatus()
+	//that listen to the user's abort command
+
+	ctxApprove, cancelApprove := context.WithTimeout(context.Background(), 10*time.Minute)
```
Comment (tsorya) [inline src/main/assisted-installer-controller/assisted_installer_main.go:70]: why not to have same context for all of them? except logs?

File: src/assisted_installer_controller/assisted_installer_controller.go [hunk 8]
```diff
@@ -89,14 +89,34 @@ func NewController(log *logrus.Logger, cfg ControllerConfig, ops ops.Ops, ic inv
...
 	return atomic.LoadUint32(&status.errCounter) > 0
 }
 
+func (status *ControllerStatus) Abort() {
+	for _, cancelFunc := range status.cancels {
+		cancelFunc()
+	}
+}
+
+func (status *ControllerStatus) AddWatch(cancelFunc context.CancelFunc) {
```
Comment (tsorya) [inline src/assisted_installer_controller/assisted_installer_controller.go:114]: i don't understand this :( why?

File: src/assisted_installer_controller/assisted_installer_controller.go [hunk 9]
```diff
@@ -241,108 +261,130 @@ func isCsrApproved(csr *certificatesv1.CertificateSigningRequest) bool {
...
+	//if the condition for post install steps is not met within the timeout
+	//or the operation is canceled, return without configuring
+	if !ok {
+		return
 	}
 
-	errMessage := ""
-	err := c.postInstallConfigs()
+	var errMessage string
+	err, canceled := c.postInstallConfigs(ctx)
```
Comment (tsorya) [inline src/assisted_installer_controller/assisted_installer_controller.go:284]: you can just verify context was cancelled or not

File: .gitignore [hunk 10]
```diff
@@ -1,3 +1,4 @@
 build
 reports
 **/RCS/**
+.vscode
```
Comment (eranco74) [inline .gitignore:4]: Please remove.
You can add it to `~/.gitignore_global`

File: src/assisted_installer_controller/assisted_installer_controller_test.go [hunk 11]
```diff
@@ -776,14 +778,14 @@ var _ = Describe("installer HostRoleMaster role", func() {
 
 	})
 
-	Context("Upload logs with oc must-gather", func() {
+	FContext("Upload logs with oc must-gather", func() {
```
Comment (eranco74) [inline src/assisted_installer_controller/assisted_installer_controller_test.go:781]: What's Fcontext?

File: src/assisted_installer_controller/assisted_installer_controller_test.go [hunk 12]
```diff
@@ -807,7 +810,7 @@ var _ = Describe("installer HostRoleMaster role", func() {
 				}).AnyTimes()
 			reportLogProgressSuccess()
 		})
-		It("Validate upload logs (with must-gather logs)", func() {
+		FIt("Validate upload logs (with must-gather logs)", func() {
```
Comment (eranco74) [inline src/assisted_installer_controller/assisted_installer_controller_test.go:813]: Is this a thing?

File: src/main/assisted-installer-controller/assisted_installer_main.go [hunk 13]
```diff
@@ -55,36 +55,48 @@ func main() {
...
-	wg.Add(1)
-	go assistedController.UpdateBMHs(&wg)
+
+	go assistedController.UpdateBMHs(bmhconfig, &wg)
 	wg.Add(1)
 
-	ctxLogs, cancelLogs := context.WithCancel(context.Background())
-	go assistedController.UploadLogs(ctxLogs, cancelLogs, &wgLogs, &status)
+	go assistedController.UploadLogs(ctxLogs, &wgLogs, status)
 	wgLogs.Add(1)
```
Comment (eranco74) [inline src/main/assisted-installer-controller/assisted_installer_main.go:86]: I think that now you don't need a special WaitGroup for the logs goroutine.

File: src/main/assisted-installer-controller/assisted_installer_main.go [hunk 14]
```diff
@@ -55,36 +55,48 @@ func main() {
...
+	wg.Add(1)
 
 	logger.Infof("Waiting for all go routines to finish")
 	wg.Wait()
-	if !status.HasError() {
-		//with error the logs are canceled within UploadLogs
-		logger.Infof("closing logs...")
-		cancelLogs()
-	}
+	// if !status.HasError() {
```
Comment (eranco74) [inline src/main/assisted-installer-controller/assisted_installer_main.go:96]: Why comment out and not delete?

**Review by eranco74 — COMMENTED** [review]
This seems like an overkill if all we want is to stop the controller in case of cancellation. 
Consider doing the refactoring in a separate PR.

**Comment by slaviered** [comment]
> This seems like an overkill if all we want is to stop the controller in case of cancellation.
> Consider doing the refactoring in a separate PR.

How would you stop the multiple wait loops in a simpler way (i.e. without passing the context to each of the goroutine and check in every iteration of the wait loop that the controller is not aborted?)

**Pull request closed without merging**